spopt/tests/test_p_dispersion.py

```diff
@@ -0,0 +1,222 @@
+        network_distance = pandas.read_csv(
+            self.dirpath
+            + "SF_network_distance_candidateStore_16_censusTract_205_new.csv"
+        )
+
+        ntw_dist_piv = network_distance.pivot_table(
+            values="distance", index="DestinationName", columns="name"
+        )
+
+        self.cost_matrix = ntw_dist_piv.to_numpy()
+
+        facility_points = pandas.read_csv(self.dirpath + "SF_store_site_16_longlat.csv")
+
+        self.facility_points_gdf = (
+            geopandas.GeoDataFrame(
+                facility_points,
+                geometry=geopandas.points_from_xy(
+                    facility_points.long, facility_points.lat
+                ),
+            )
+            .sort_values(by=["NAME"])
+            .reset_index()
+        )
+
+        self.service_dist = 5000.0
+        self.p_facility = 4
+
+
+    def test_optimality_p_dispersion_from_cost_matrix(self):
+        pdispersion = PDispersion.from_cost_matrix(
+            self.cost_matrix, p_fac=self.p_facility
+        )
+        pdispersion = pdispersion.solve(pulp.PULP_CBC_CMD(msg=False))
+        self.assertEqual(pdispersion.problem.status, pulp.LpStatusOptimal)
+    # how to force this to fail???
+    def test_infeasibility_p_dispersion_from_cost_matrix(self):
+        pdispersion = PDispersion.from_cost_matrix(self.cost_matrix, p_fac=0)
+        with self.assertRaises(RuntimeError):
+            pdispersion.solve(pulp.PULP_CBC_CMD(msg=False))
+
+    def test_optimality_p_dispersion_from_geodataframe(self):
+        pdispersion = PDispersion.from_geodataframe(
+            self.facility_points_gdf, "geometry", p_fac=self.p_facility,
+        )
+        pdispersion = pdispersion.solve(pulp.PULP_CBC_CMD(msg=False))
+        self.assertEqual(pdispersion.problem.status, pulp.LpStatusOptimal)
+    # how to force this to fail??? a cost matrix with all facilities equidistant?
+    def test_infeasibility_p_dispersion_from_geodataframe(self):
+        pdispersion = PDispersion.from_geodataframe(
+            self.facility_points_gdf, "geometry", p_fac=0,
+        )
+        with self.assertRaises(RuntimeError):
+            pdispersion.solve(pulp.PULP_CBC_CMD(msg=False))
+
+class TestErrorsWarnings(unittest.TestCase):
+    def setUp(self) -> None:
+
+        pol1 = Polygon([(0, 0), (1, 0), (1, 1)])
+        pol2 = Polygon([(0, 0), (1, 0), (1, 1), (0, 1)])
+        pol3 = Polygon([(2, 0), (3, 0), (3, 1), (2, 1)])
+        polygon_dict = {"geometry": [pol1, pol2, pol3]}
+
+        point = Point(10, 10)
+        point_dict = {"weight": 4, "geometry": [point]}
+
+        self.gdf_fac = geopandas.GeoDataFrame(polygon_dict, crs="EPSG:4326")
+        self.gdf_dem = geopandas.GeoDataFrame(point_dict, crs="EPSG:4326")
+
+        self.gdf_dem_crs = self.gdf_dem.to_crs("EPSG:3857")
+
+        self.gdf_dem_buffered = self.gdf_dem.copy()
+        self.gdf_dem_buffered["geometry"] = self.gdf_dem.buffer(2)
+
+    # i don't think this specific test is relevant
```

Comment:
Let's include the test to be thorough.

Reply:
This test is checking for a different crs between two geodataframes, but p-dispersion only takes a single geodataframe.

Reply:
@erinrolson aha! good point!